File: frontend-v2/src/shared/api/inbox.ts
// Message-centre transport (docs/MESSAGE_CENTER.md §用户接口). Query keys carry
// the user id (§7.2); `inbox.updated` on the socket invalidates the lot.
import { useEffect } from "react"
import { useInfiniteQuery, useQuery, useQueryClient } from "@tanstack/react-query"
import { http } from "./http"
import { useAuthStore } from "./auth-store"
import { wsClient } from "../ws/client"

export type InboxCategory = "session" | "system" | "notice"
export const INBOX_CATEGORIES: readonly InboxCategory[] = ["session", "system", "notice"]

/** Allow-listed navigation targets; anything else opens the inbox itself. */
export type InboxLink =
  | { kind: "session"; workspaceId: string; sessionId: string; panel?: "desktop"; control?: boolean }
  | { kind: "cron"; workspaceId: string; jobId?: string }
  | { kind: "auth_center"; workspaceId: string; jobId?: string }
  | { kind: "skills"; workspaceId: string }
  | { kind: "admin_skills" }
  | { kind: "topic"; slug: string }
  | { kind: "url"; url: string }
  | { kind: "inbox" }
  | { kind: string }

export interface InboxItem {
  id: string
  category: InboxCategory
  kind: string
  title: string
  body: string
  link: InboxLink | null
  workspaceId: string | null
  announcementId: string | null
  readAt: string | null
  resolvedAt: string | null
  expiresAt: string | null
  createdAt: string
}

export interface InboxUnread {
  total: number
  session: number
  system: number
  notice: number
}

export interface InboxPage {
  items: InboxItem[]
  nextCursor: string | null
  unread: InboxUnread
}

export interface TopicPage {
  id: string
  slug: string
  title: string
  coverUrl: string | null
  contentMd: string
  ctaLabel: string | null
  ctaLink: InboxLink | null
  publishedAt: string | null
  updatedAt: string
}

export const inboxKeys = {
  all: (userId: string) => ["inbox", userId] as const,
  unread: (userId: string) => ["inbox", userId, "unread"] as const,
  feed: (userId: string, category: InboxCategory | "") => ["inbox", userId, "feed", category] as const,
  topic: (slug: string) => ["topic", slug] as const,
}

export function useInboxUnread() {
  const userId = useAuthStore((s) => s.user?.id ?? "anonymous")
  const enabled = useAuthStore((s) => s.isAuthenticated)
  return useQuery({
    queryKey: inboxKeys.unread(userId),
    enabled,
    queryFn: ({ signal }) => http.get<InboxUnread>("/api/inbox/unread", { signal }),
    refetchInterval: 120_000,
  })
}

export function useInboxFeed(category: InboxCategory | "") {
  const userId = useAuthStore((s) => s.user?.id ?? "anonymous")
  return useInfiniteQuery({
    queryKey: inboxKeys.feed(userId, category),
    initialPageParam: null as string | null,
    queryFn: ({ pageParam, signal }) => {
      const params = new URLSearchParams({ limit: "30" })
      if (category) params.set("category", category)
      if (pageParam) params.set("cursor", pageParam)
      return http.get<InboxPage>(`/api/inbox?${params}`, { signal })
    },
    getNextPageParam: (last) => last.nextCursor,
  })
}

/** Idempotent; the response is the row itself, link included. */
export const markRead = (id: string) => http.post<InboxItem>(`/api/inbox/${encodeURIComponent(id)}/read`)

export const readAll = (category: InboxCategory | "") =>
  http.post<{ updated: number }>(`/api/inbox/read-all${category ? `?category=${category}` : ""}`)

/** Public: no bearer needed, drafts are 404. */
export function useTopic(slug: string) {
  return useQuery({
    queryKey: inboxKeys.topic(slug),
    enabled: slug.length > 0,
    retry: false,
    queryFn: ({ signal }) => http.get<TopicPage>(`/api/topics/${encodeURIComponent(slug)}`, { signal }),
  })
}

/** Mount once under the workspace layout: badge and open feeds follow the socket. */
export function useInboxLiveEvents(): void {
  const qc = useQueryClient()
  const userId = useAuthStore((s) => s.user?.id ?? "anonymous")
  useEffect(() => {
    const invalidate = () => void qc.invalidateQueries({ queryKey: inboxKeys.all(userId) })
    const offs = [wsClient.on("inbox.updated", invalidate), wsClient.on("__connected", invalidate)]
    return () => offs.forEach((off) => off())
  }, [qc, userId])
}

File: mobile/lib/shared/models/inbox.dart
import 'json.dart';

/// Message-centre models (backend `notifications/inbox.py::public_item`,
/// `api/inbox.py::public_topic`; see docs/MESSAGE_CENTER.md).

/// Inbox tabs, in nav order. `null` category means "all".
const inboxCategories = ['session', 'system', 'notice'];

/// An allow-listed navigation target. Anything the app does not recognise
/// resolves to the inbox itself — never to a URL the server did not vet.
class InboxLink {
  const InboxLink({
    required this.kind,
    this.workspaceId,
    this.sessionId,
    this.jobId,
    this.slug,
    this.url,
    this.panel,
    this.control = false,
  });

  static InboxLink? fromJson(dynamic json) {
    if (json is! Map<String, dynamic>) return null;
    final kind = asString(json['kind']);
    if (kind == null || kind.isEmpty) return null;
    return InboxLink(
      kind: kind,
      workspaceId: asString(json['workspaceId']),
      sessionId: asString(json['sessionId']),
      jobId: asString(json['jobId']),
      slug: asString(json['slug']),
      url: asString(json['url']),
      panel: asString(json['panel']),
      control: asBool(json['control']) ?? false,
    );
  }

  final String kind;
  final String? workspaceId;
  final String? sessionId;
  final String? jobId;
  final String? slug;
  final String? url;
  final String? panel;
  final bool control;
}

class InboxItem {
  const InboxItem({
    required this.id,
    required this.category,
    required this.kind,
    required this.title,
    required this.body,
    required this.createdAt,
    this.link,
    this.workspaceId,
    this.announcementId,
    this.readAt,
    this.resolvedAt,
  });

  factory InboxItem.fromJson(Map<String, dynamic> json) => InboxItem(
    id: asString(json['id']) ?? '',
    category: asString(json['category']) ?? 'system',
    kind: asString(json['kind']) ?? '',
    title: asString(json['title']) ?? '',
    body: asString(json['body']) ?? '',
    link: InboxLink.fromJson(json['link']),
    workspaceId: asString(json['workspaceId']),
    announcementId: asString(json['announcementId']),
    readAt: asDate(json['readAt']),
    resolvedAt: asDate(json['resolvedAt']),
    createdAt: asDate(json['createdAt']) ?? DateTime.now(),
  );

  final String id;
  final String category;
  final String kind;
  final String title;
  final String body;
  final InboxLink? link;
  final String? workspaceId;
  final String? announcementId;
  final DateTime? readAt;
  final DateTime? resolvedAt;
  final DateTime createdAt;

  bool get unread => readAt == null;

  InboxItem asRead() => InboxItem(
    id: id,
    category: category,
    kind: kind,
    title: title,
    body: body,
    link: link,
    workspaceId: workspaceId,
    announcementId: announcementId,
    readAt: readAt ?? DateTime.now(),
    resolvedAt: resolvedAt,
    createdAt: createdAt,
  );
}

class InboxUnread {
  const InboxUnread({
    this.total = 0,
    this.session = 0,
    this.system = 0,
    this.notice = 0,
  });

  factory InboxUnread.fromJson(Map<String, dynamic> json) => InboxUnread(
    total: asInt(json['total']) ?? 0,
    session: asInt(json['session']) ?? 0,
    system: asInt(json['system']) ?? 0,
    notice: asInt(json['notice']) ?? 0,
  );

  final int total;
  final int session;
  final int system;
  final int notice;

  int forCategory(String? category) => switch (category) {
    'session' => session,
    'system' => system,
    'notice' => notice,
    _ => total,
  };
}

class InboxPage {
  const InboxPage({
    required this.items,
    this.nextCursor,
    this.unread = const InboxUnread(),
  });

  factory InboxPage.fromJson(Map<String, dynamic> json) => InboxPage(
    items: asList(
      json['items'],
    ).whereType<Map<String, dynamic>>().map(InboxItem.fromJson).toList(),
    nextCursor: asString(json['nextCursor']),
    unread: InboxUnread.fromJson(asMap(json['unread'])),
  );

  final List<InboxItem> items;
  final String? nextCursor;
  final InboxUnread unread;
}

class TopicPage {
  const TopicPage({
    required this.slug,
    required this.title,
    required this.contentMd,
    this.coverUrl,
    this.ctaLabel,
    this.ctaLink,
    this.publishedAt,
  });

  factory TopicPage.fromJson(Map<String, dynamic> json) => TopicPage(
    slug: asString(json['slug']) ?? '',
    title: asString(json['title']) ?? '',
    contentMd: asString(json['contentMd']) ?? '',
    coverUrl: asString(json['coverUrl']),
    ctaLabel: asString(json['ctaLabel']),
    ctaLink: InboxLink.fromJson(json['ctaLink']),
    publishedAt: asDate(json['publishedAt']),
  );

  final String slug;
  final String title;
  final String contentMd;
  final String? coverUrl;
  final String? ctaLabel;
  final InboxLink? ctaLink;
  final DateTime? publishedAt;
}
